librad/src/git/repo.rs

```diff
@@ -0,0 +1,589 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    fs,
+};
+
+use serde::{de::DeserializeOwned, Serialize};
+use thiserror::Error;
+
+use crate::{
+    canonical::{Cjson, CjsonError},
+    git::{
+        refs::{self, Oid, Refs},
+        url::GitUrlRef,
+    },
+    hash::{self, Hash},
+    keys::device,
+    meta::entity::{
+        self,
+        data::{EntityBuilder, EntityData},
+        Entity,
+        Signatory,
+        VerificationStatus,
+    },
+    paths::Paths,
+    peer::PeerId,
+    uri::{self, RadUrl, RadUrn},
+};
+
+const RAD_REFS: &str = "rad/refs";
+
+pub struct Repo {
+    id: Hash,
+    key: device::Key,
+    repo: git2::Repository,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum Error {
+    #[error("Repo already exists")]
+    AlreadyExists,
+
+    #[error("Unknown repo")]
+    NoSuchRepo,
+
+    #[error("Branch {0} not found")]
+    NoSuchBranch(String),
+
+    #[error("Blob {0} not found")]
+    NoSuchBlob(String),
+
+    #[error("Identity VerificationStatus must be {expected:?}, got {actual:?}")]
+    Verification {
+        expected: VerificationStatus,
+        actual: VerificationStatus,
+    },
+
+    #[error(
+        "Identity root hash doesn't match resolved URL. Expected {expected}, actual: {actual}"
+    )]
+    RootHashMismatch { expected: Hash, actual: Hash },
+
+    #[error(transparent)]
+    HashAlgorithm(#[from] hash::AlgorithmMismatch),
+
+    #[error(transparent)]
+    Entity(#[from] entity::Error),
+
+    #[error(transparent)]
+    Refsig(#[from] refs::signed::Error),
+
+    #[error(transparent)]
+    Json(#[from] serde_json::error::Error),
+
+    #[error(transparent)]
+    Cjson(#[from] CjsonError),
+
+    #[error(transparent)]
+    Git(#[from] git2::Error),
+}
+
+impl Repo {
+    pub fn create<T>(paths: &Paths, key: device::Key, meta: &Entity<T>) -> Result<Self, Error>
```

Comment:
To check I'm understanding correctly: this allows you to create a repo for any entity, whether it's a project or a user? And this becomes tracking information in the current project you're working on?

Reply:
Yup. As I said in RTC, I'm not fully convinced that knowning `T` for `clone` is always practical -- the caller may have just a URN, so it might not be clear from the context what `T` should be. We can fix that by having a value-level bit in the URN which tells us what `T` is.